### ISSUE TITLE
Appropriate way to use spec-cluster with dask-cuda

### ISSUE THREAD

**Issue by pentschev** [rapidsai/dask-cuda#146]
Currently dask-cuda has no class similar to distributed's `Worker`, given it currently doesn't need a specialized worker. However, we have a CLI that extends `dask-worker`, but it can't be directly used with spec-cluster as we may need additional parameters to pass to each process of `dask-cuda-worker`.

Recently, a new `dask-spec-worker` was proposed https://github.com/dask/distributed/pull/3090 which seems like a good way for enabling additional functionality to `dask-cuda-worker`, but as far as I understand it, we cannot directly pass the CLI interface with that, we would require some `CUDAWorker` class for that.

@mrocklin any suggestions on what would be an appropriate solution here?

**Comment by mrocklin** [comment]
> Currently dask-cuda has no class similar to distributed's Worker, given it currently doesn't need a specialized worker

Yes, this is, I think, by design.  Ideally we don't need to create a specialized `Worker` class.  I think that this is best in the long run for maintenance reasons.  The fact that we had to have a `dask-cuda-worker` is, I hope, a short term workaround.

We use the current `Nanny` class with `SpecCluster` today in `DGX`.  I think that it works well.

So, to the questions asked in this issue my initial gut reaction is "we're already done".  However I suspect that there is some deeper issue that you're trying to resolve.  Can I ask what caused you to raise this issue?

I think that there are a few different questions here:

**Comment by pentschev** [comment]
Additionally, we don't want to force users to specify the options that are already available with `dask-cuda-worker`, even though it's technically possible that the user mimics it just with a regular `Worker` instance (or `Nanny`, perhaps).

**Comment by mrocklin** [comment]
> Is that your main proposal?

I don't really have a proposal.  I'm not sure what the best course of action here is yet.  Thoughts are welcome.

**Comment by mrocklin** [comment]
I think that having "please do everything right automatically for GPUs" user options makes sense.  I would like to reduce the amount of forks in the codebase if possible, just to reduce friction when we update things upstream.  Designing things well to achieve both of these goals simultaneously is hard :/

**Comment by mrocklin** [comment]
I think that we're building out some functionality that will be useful in the final product, like `SpecCluster`, your automatic determination of CPU affinity, GPUs, and so on.  I don't know the right way to assemble everything though.

**Comment by pentschev** [comment]
I think having things based on spec-cluster is generally a good idea. Admittedly, I don't know it in details, so I think it would be useful to have some way of launching workers with all CUDA capabilities alone (as `dask-cuda-worker` does). Not sure if this is a thing already, but maybe it would be interesting to have some sort of recipe for workers? In the case here, maybe we could have a way of creating this particular `dict` https://github.com/rapidsai/dask-cuda/blob/branch-0.10/dask_cuda/dgx.py#L58-L81 based on predefined settings and parameters passed by the user.

Right now, I need a more general way of starting workers that may rely on some information to setup environment variables to enable UCX functionality, and I have to be able to start these workers on the local machine or on remote machine(s) with `SSHCluster`.

I'm also not sure what would be a good solution, but perhaps a function/class that allows passing some information to assemble the spec dictionary would be a simple way out.

**Comment by mrocklin** [comment]
> but perhaps a function/class that allows passing some information to assemble the spec dictionary would be a simple way out.

That sounds like a good idea to me.  I think that such a function would be useful no matter what the final solution looks like.  Presumably it could also include the various automatic detection options that we have.

Even if the `dask-spec-worker` CLI isn't useful in this case (because we need to make a specific `dask-cuda-worker`, maybe the approach used in `dask-spec-worker` could be used to simplify `dask-cuda-worker`.

**Comment by jakirkham** [comment]
cc @jacobtomlinson (in case this is of interest 🙂)

**Comment by jacobtomlinson** [comment]
Thanks @jakirkham.

I implemented a `CUDAWorker` in #324. But it isn't a direct replacement for `Worker`. The `CUDAWorker` will create an internal `Worker` instance for each GPU and configure it correctly. So you only need one `CUDAWorker` per node.

This makes sense with `SSHCluster` as each host you SSH to should get one worker per GPU. But for `SpecCluster` you deal more in terms of workers.

We may want to factor things out further, I did try that initially but things got complex in terms of communicating between workers at startup to decide which GPU would be assigned to each.
